### PR TITLE
Conflict spike1

### PR DESCRIPTION
**Pull request by Lavanyagaur22**
### Conflict Resolution Spike 1 
(Approached Version based conflicts resolution according to our schema)

#### What is done
1. Using the ConflictResolution Interceptor attached to the OkHttpClient of the apollo client, the response from the server is read and if the response contains the string- "Voyager Conflict", we parse the json response (errors) to a class ConflictPojo like this:

Screenshot 2019-07-20 at 4 23 49 AM

2. In case of conflicts, data returned by the server on performing a mutation is null. So taking this advantage and checking the body of the response.data in the callback (onResponse) of the particular mutation, if the response body is null, we have done the following:

- Get access to the object of ConflictPojo class.
- Retrieve the server version of the above object.
- Make a mutation call with the correct version (rest of the things being the same).

Screenshot 2019-07-20 at 4 29 19 AM

#### Problem To Be Fixed
While showing toast, we need activity context which we currently don't have access to and we can't cast application to activity context as it's showing class cast exception. (have to be fixed)

### CODE REVIEW

File: sample/src/main/java/org/aerogear/graphqlandroid/ConflictPojo.kt [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+)
+class ExceptionClass(
+    val conflictInfo : ConflictData
+)
+
+class ConflictData(
+    val serverState : ServerState
+)
+
+class ServerState(
```
Comment (wtrocki) [inline sample/src/main/java/org/aerogear/graphqlandroid/ConflictPojo.kt:22]: Do people need to define pojo in their app fr each type? Why this way? any reason we need this.

File: sample/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -5,7 +5,7 @@
     <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE"/>
 
     <application
-            android:name=".MyApplciation"
+            android:name=".MyApplication"
```
Comment (wtrocki) [inline sample/src/main/AndroidManifest.xml:8]: OffixExample? Use fox as icon

File: sample/src/main/java/org/aerogear/graphqlandroid/ConflictPojo.kt [hunk 3]
```diff
@@ -0,0 +1,28 @@
+package org.aerogear.graphqlandroid
```
Comment (wtrocki) [inline sample/src/main/java/org/aerogear/graphqlandroid/ConflictPojo.kt:1]: IMHO this class is not needed. devs would not see conflict parts in their code

File: sample/src/main/java/org/aerogear/graphqlandroid/Utils.kt [hunk 4]
```diff
@@ -154,6 +163,14 @@ object Utils {
         return conflictInterceptor
     }
 
+    /*
+     This function parses the response body of the conflict detected and makes an object of ConflictPojo class.
+     */
+    fun makePojoOfResponse(responseBodyString: String) {
+        parsedObject = gson.fromJson(responseBodyString, ConflictPojo::class.java)
```
Comment (wtrocki) [inline sample/src/main/java/org/aerogear/graphqlandroid/Utils.kt:170]: Any reason we need that inside example app?

**Review by wtrocki — CHANGES_REQUESTED** [review]
Code should be in library. Please describe how you will see end user api done in description.

**Comment by Lavanyagaur22** [comment]
@wtrocki This is a sample case done in the app to check whether this approach works or not. Would be implementing similar thing in the library.

File: sample/src/main/java/org/aerogear/graphqlandroid/ConflictPojo.kt (hunk 1 above)
Comment (Lavanyagaur22) [reply sample/src/main/java/org/aerogear/graphqlandroid/ConflictPojo.kt:22]: Not in their app, we would be having it in the library.

File: sample/src/main/java/org/aerogear/graphqlandroid/ConflictPojo.kt (hunk 3 above)
Comment (Lavanyagaur22) [reply sample/src/main/java/org/aerogear/graphqlandroid/ConflictPojo.kt:1]: This would be there in the library.

**Comment by Lavanyagaur22** [comment]
Btw, we are checking conflicts by checking the "voyager conflict" string in the error interceptor. This is here schema specific. 

Can we say that when we perform any mutation and receive null data, this is the case of conflict?

**Comment by wtrocki** [comment]
> Can we say that when we perform any mutation and receive null data, this is the case of conflict?

Null data can mean may different problems that will be usually listed in errors array. What is the problem with detecting conflict error? Can you elaborate little bit more?

**Comment by wtrocki** [comment]
Let's not merge that until we have this in the library. I do not see this code working in the library due to the fact that we are using business object pojos instead of hashmaps

**Comment by harshithdwivedi** [comment]
Closing this in favour of #94

**Pull request closed without merging**